File: search/graph-tree-search/breadth-first-search/python/breadth_first_search.py
from collections import deque
from typing import List, Dict, Set, Optional

def bfs_graph(
    graph: Dict[str, List[str]], 
    start: str, 
    target: Optional[str] = None
) -> List[str]:
    """
    Perform Breadth-First Search on a graph represented as an adjacency list.
    
    Args:
        graph (Dict[str, List[str]]): Adjacency list representation of the graph.
                                      Example: {'A': ['B', 'C'], 'B': ['D'], ...}
        start (str): Starting node for traversal.
        target (Optional[str]): Optional target node to search for.
    
    Returns:
        List[str]: Order of visited nodes. If target is provided and found,
                   returns path from start to target (if path reconstruction is added).
    """
    if start not in graph:
        raise ValueError(f"Start node '{start}' not in graph.")
    
    # Initialize queue with the start node
    queue = deque([start])
    visited: Set[str] = {start}
    traversal_order: List[str] = [start]
    
    while queue:
        current = queue.popleft()
        
        # If searching for a specific target
        if target and current == target:
            print(f"🎯 Target '{target}' found!")
            return traversal_order  
        
        # Explore all unvisited neighbours
        for neighbour in graph.get(current, []):
            if neighbour not in visited:
                visited.add(neighbour)
                queue.append(neighbour)
                traversal_order.append(neighbour)
    
    return traversal_order


# 🌲 BFS for Binary Tree (Node-based structure)
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def bfs_tree(root: Optional[TreeNode]) -> List[int]:
    """
    Level-order traversal of a binary tree using BFS.
    
    Args:
        root (TreeNode or None): Root of the binary tree.
    
    Returns:
        List[int]: Values of nodes visited in level order.
    """
    if not root:
        return []
    
    queue = deque([root])
    result = []
    
    while queue:
        node = queue.popleft()
        result.append(node.val)
        
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
    
    return result


# 🧪 Example Usage & Testing
if __name__ == "__main__":
    # Example 1: Undirected Graph
    graph = {
        'A': ['B', 'C'],
        'B': ['A', 'D'],
        'C': ['A', 'E', 'F'],
        'D': ['B'],
        'E': ['C'],
        'F': ['C']
    }
    
    print("🌐 BFS on Graph:")
    order = bfs_graph(graph, start='A')
    print("Traversal order:", " → ".join(order))
    # Output: A → B → C → D → E → F
    
    print("\n🔍 Searching for target 'E':")
    bfs_graph(graph, start='A', target='E')
    
    # Example 2: Binary Tree BFS (Level-order)
    #       1
    #      / \
    #     2   3
    #    /   / \
    #   4   5   6
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    
    print("\n🌳 BFS on Binary Tree (Level-order):")
    tree_order = bfs_tree(root)
    print("Level-order:", tree_order)
    # Output: [1, 2, 3, 4, 5, 6]

File: search/graph-tree-search/breadth-first-search/javascript/breadth_first_search.js
/**
 * Perform Breadth-First Search on a graph represented as an adjacency list.
 * 
 * @param {Object.<string, string[]>} graph - Adjacency list representation of the graph.
 *                                            Example: { A: ['B', 'C'], B: ['D'], ... }
 * @param {string} start - Starting node for traversal.
 * @param {string} [target] - Optional target node to search for.
 * @returns {string[]} Order of visited nodes.
 */
function bfsGraph(graph, start, target = null) {
    if (!(start in graph)) {
        throw new Error(`Start node '${start}' not in graph.`);
    }

    // Initialize queue with the start node
    const queue = [start];
    const visited = new Set([start]);
    const traversalOrder = [start];
    let index = 0; // pointer to front of queue

    
    while (index < queue.length) {
        const current = queue[index++];

        // If searching for a specific target
        if (target && current === target) {
            console.log(`🎯 Target '${target}' found!`);
            return traversalOrder;
        }

        // Explore all unvisited neighbours
        const neighbours = graph[current] || [];
        for (const neighbour of neighbours) {
            if (!visited.has(neighbour)) {
                visited.add(neighbour);
                queue.push(neighbour); // enqueue at end
                traversalOrder.push(neighbour);
            }
        }
    }

    return traversalOrder;
}

// 🌲 BFS for Binary Tree (Node-based structure)
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/**
 * Level-order traversal of a binary tree using BFS.
 * 
 * @param {TreeNode | null} root - Root of the binary tree.
 * @returns {number[]} Values of nodes visited in level order.
 */
function bfsTree(root) {
    if (!root) return [];

    const queue = [root];
    const result = [];

    while (queue.length > 0) {
        const node = queue.shift();
        result.push(node.val);

        if (node.left) queue.push(node.left);
        if (node.right) queue.push(node.right);
    }

    return result;
}

// 🧪 Example Usage & Testing
(() => {
    // Example 1: Undirected Graph
    const graph = {
        'A': ['B', 'C'],
        'B': ['A', 'D'],
        'C': ['A', 'E', 'F'],
        'D': ['B'],
        'E': ['C'],
        'F': ['C']
    };

    console.log("🌐 BFS on Graph:");
    const order = bfsGraph(graph, 'A');
    console.log("Traversal order:", order.join(" → "));
    // Output: A → B → C → D → E → F

    console.log("\n🔍 Searching for target 'E':");
    bfsGraph(graph, 'A', 'E');

    // Example 2: Binary Tree BFS (Level-order)
    //       1
    //      / \
    //     2   3
    //    /   / \
    //   4   5   6
    const root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.left = new TreeNode(4);
    root.right.left = new TreeNode(5);
    root.right.right = new TreeNode(6);

    console.log("\n🌳 BFS on Binary Tree (Level-order):");
    const treeOrder = bfsTree(root);
    console.log("Level-order:", treeOrder);
    // Output: [1, 2, 3, 4, 5, 6]
})();